# Celebal Technologies Internship Assignment
## Python Basics — Data Exploration & Cleaning with Pandas

**Dataset:** [Superstore Sales Dataset (Kaggle)](https://www.kaggle.com/datasets/vivek468/superstore-dataset-final)
**File used:** `Sample - Superstore.csv`

**Objective:** Load, explore, clean, and transform the Superstore dataset using Pandas, then save the cleaned output as a new CSV file.

**Steps covered:**
1. Load CSV into a DataFrame
2. Explore data (head/tail, shape, columns, dtypes)
3. Handle missing values
4. Basic operations (filter rows, select columns)
5. Remove duplicates
6. Create derived column (`total_amount = Unit_Price * Quantity`)
7. Save cleaned dataset as new CSV


## Step 0: Import Libraries

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)


## Step 1: Load the CSV Dataset

The Superstore CSV is loaded into a Pandas DataFrame. This dataset is typically saved with
`latin1`/`ISO-8859-1` encoding (not UTF-8), so we specify that explicitly to avoid a
`UnicodeDecodeError`.


In [2]:
df = pd.read_csv("Sample - Superstore.csv", encoding="latin1")
print("Dataset loaded successfully!")
print(f"Shape: {df.shape}")


Dataset loaded successfully!
Shape: (9994, 21)


## Step 2: Explore the Data

Checking the first/last rows, shape, column names, data types, and summary statistics
to understand the dataset's structure.


In [3]:
# First 5 rows
df.head()


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [4]:
# Last 5 rows
df.tail()


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,Florida,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.248,3,0.2,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.960,2,0.0,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.576,2,0.2,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.600,4,0.0,13.3200
9993,9994,CA-2017-119914,5/4/2017,5/9/2017,Second Class,CC-12220,Chris Cortes,Consumer,United States,Westminster,California,92683,West,OFF-AP-10002684,Office Supplies,Appliances,"Acco 7-Outlet Masterpiece Power Center, Wihtou...",243.160,2,0.0,72.9480


In [5]:
# Shape of the dataset (rows, columns)
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")


Rows: 9994, Columns: 21


In [6]:
# Column names
df.columns.tolist()


['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit']

In [7]:
# Data types of each column
df.dtypes


Row ID             int64
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code        int64
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object

In [8]:
# Summary statistics for numeric columns
df.describe()


,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


In [9]:
# General info (dtypes + non-null counts)
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

## Step 3: Handle Missing Values

**Identify:** Check how many missing values each column has.

**Strategy (applied generically, in case any column has nulls):**
- Numeric columns (`Sales`, `Quantity`, `Discount`, `Profit`, `Postal Code`) → fill with the **median**.
- Categorical columns (`Ship Mode`, `Region`, `State`, `City`) → fill with the **mode**.

This is the Kaggle "cleaned" version of the Superstore dataset, so it is expected to already
have zero missing values — but the code below is written to handle them generically so it
still works correctly on any messier version of the dataset.


In [10]:
# Identify missing values
missing_before = df.isnull().sum()
print("Missing values per column:")
print(missing_before[missing_before > 0] if missing_before.sum() > 0 else "No missing values found in the dataset.")


Missing values per column:
No missing values found in the dataset.


In [11]:
# Generic handling: fill numeric columns with median, categorical columns with mode
numeric_cols = df.select_dtypes(include=[np.number]).columns
categorical_cols = df.select_dtypes(include=['object']).columns

for col in numeric_cols:
    if df[col].isnull().sum() > 0:
        df[col] = df[col].fillna(df[col].median())

for col in categorical_cols:
    if df[col].isnull().sum() > 0:
        df[col] = df[col].fillna(df[col].mode()[0])

print("Missing values after cleaning:", df.isnull().sum().sum())


Missing values after cleaning: 0


/tmp/ipykernel_561/2541886636.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include=['object']).columns


## Step 4: Basic Operations (Filter Rows, Select Columns)

Examples of common Pandas operations on the Superstore data.


In [12]:
# Select specific columns
subset = df[['Order ID', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Profit']]
subset.head()


,Order ID,Category,Sub-Category,Sales,Quantity,Profit
0,CA-2016-152156,Furniture,Bookcases,261.9600,2,41.9136
1,CA-2016-152156,Furniture,Chairs,731.9400,3,219.5820
2,CA-2016-138688,Office Supplies,Labels,14.6200,2,6.8714
3,US-2015-108966,Furniture,Tables,957.5775,5,-383.0310
4,US-2015-108966,Office Supplies,Storage,22.3680,2,2.5164


In [13]:
# Filter rows: Technology orders with Sales > 500
tech_high_value = df[(df['Category'] == 'Technology') & (df['Sales'] > 500)]
print(f"Technology orders with Sales > 500: {len(tech_high_value)}")
tech_high_value.head()


Technology orders with Sales > 500: 396


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
7,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.152,6,0.2,90.7152
11,12,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,TEC-PH-10002033,Technology,Phones,Konftel 250 Conference phone - Charcoal black,911.424,4,0.2,68.3568
35,36,CA-2016-117590,12/8/2016,12/10/2016,First Class,GH-14485,Gene Hale,Corporate,United States,Richardson,Texas,75080,Central,TEC-PH-10004977,Technology,Phones,GE 30524EE4,1097.544,7,0.2,123.4737
54,55,CA-2016-105816,12/11/2016,12/17/2016,Standard Class,JM-15265,Janet Molinari,Corporate,United States,New York City,New York,10024,East,TEC-PH-10002447,Technology,Phones,AT&T CL83451 4-Handset Telephone,1029.950,5,0.0,298.6855
159,160,CA-2016-114104,11/20/2016,11/24/2016,Standard Class,NP-18670,Nora Paige,Consumer,United States,Edmond,Oklahoma,73034,Central,TEC-PH-10004536,Technology,Phones,Avaya 5420 Digital phone,944.930,7,0.0,236.2325


In [14]:
# Filter rows: loss-making orders (Profit < 0)
loss_orders = df[df['Profit'] < 0]
print(f"Orders that resulted in a loss: {len(loss_orders)}")


Orders that resulted in a loss: 1871


In [15]:
# Filter rows: orders from the West region
west_orders = df[df['Region'] == 'West']
print(f"Orders from West region: {len(west_orders)}")


Orders from West region: 3203


## Step 5: Remove Duplicates

Check for and remove any fully duplicated rows in the dataset.


In [16]:
# Check duplicates
duplicate_count = df.duplicated().sum()
print(f"Number of duplicate rows found: {duplicate_count}")


Number of duplicate rows found: 0


In [17]:
# Remove duplicates (safe to run even if none exist)
df = df.drop_duplicates().reset_index(drop=True)
print(f"Shape after removing duplicates: {df.shape}")


Shape after removing duplicates: (9994, 21)


## Step 6: Create a Derived Column

The Superstore dataset provides `Sales` as the total line-item revenue (i.e. price × quantity
already multiplied together) rather than a separate unit price column. To follow the
assignment's `total_amount = price * quantity` requirement in a meaningful way, we first
derive a `Unit_Price` column, then recompute `total_amount` from it — which also serves as a
validation check against the original `Sales` column.


In [18]:
df['Unit_Price'] = round(df['Sales'] / df['Quantity'], 2)
df['total_amount'] = round(df['Unit_Price'] * df['Quantity'], 2)

df[['Sales', 'Quantity', 'Unit_Price', 'total_amount']].head()


,Sales,Quantity,Unit_Price,total_amount
0,261.9600,2,130.98,261.96
1,731.9400,3,243.98,731.94
2,14.6200,2,7.31,14.62
3,957.5775,5,191.52,957.60
4,22.3680,2,11.18,22.36


In [19]:
# Sanity check: total_amount should match the original Sales column (small rounding differences allowed)
diff = (df['total_amount'] - df['Sales']).abs()
print(f"Max difference between total_amount and Sales: {diff.max():.4f}")


Max difference between total_amount and Sales: 0.0560


## Step 7: Save the Cleaned Dataset

Export the cleaned DataFrame to a new CSV file.


In [20]:
df.to_csv("superstore_cleaned.csv", index=False)
print("Cleaned dataset saved as 'superstore_cleaned.csv'")
print(f"Final shape: {df.shape}")


Cleaned dataset saved as 'superstore_cleaned.csv'
Final shape: (9994, 23)


## Summary

| Step | Action | Result |
|------|--------|--------|
| Load | Read raw CSV (`latin1` encoding) | 9,994 rows, 21 columns loaded |
| Explore | head/tail/shape/dtypes/describe/info | Mixed numeric + categorical columns identified |
| Missing Values | Checked all columns; generic median/mode fill logic applied | 0 missing values (dataset was already clean) |
| Filter/Select | Filtered by Category, Profit, Region; selected column subsets | Demonstrated boolean indexing & column selection |
| Duplicates | Checked & dropped exact duplicate rows | 0 duplicates found |
| Derived Column | `Unit_Price = Sales / Quantity`, `total_amount = Unit_Price * Quantity` | New columns added, validated against `Sales` |
| Save | Exported to CSV | `superstore_cleaned.csv` created (9,994 rows, 23 columns) |

**Key Pandas concepts practiced:** `read_csv` (with encoding handling), `head`/`tail`, `shape`,
`columns`, `dtypes`, `describe`, `info`, `isnull`, `fillna`, `mode`/`median`, boolean filtering,
column selection, `duplicated`, `drop_duplicates`, derived column creation, `to_csv`.
